### ISSUE TITLE
Support OpenID Connect Tokens for Kube Auth

### ISSUE THREAD

**Issue by frodopwns** [Azure/AKS#1767]
**What happened**:

Users would like to use the OpenID Connect functionality built into Kubernetes in order to authenticate using OIDC providers. Users are unable to set the parameters outlined in the Kubernetes documentation. https://kubernetes.io/docs/reference/access-authn-authz/authentication/#openid-connect-tokens

Users submitted feedback seeking resolution:

- https://github.com/Azure/AKS/issues/10
- https://feedback.azure.com/forums/914020-azure-kubernetes-service-aks/suggestions/35100826-support-oidc-parameters

No updates and the issue was closed.

**What you expected to happen**:

Issues addressed either by solving the problem or an explanation of why it can't be done or when to expect it.

**Comment by TomGeske** [comment]
You can use AKS-managed AAD https://aka.ms/aks/managed-aad

**Comment by reallydontask** [comment]
@TomGeske 

> You can use AKS-managed AAD https://aka.ms/aks/managed-aad

I am, presumably, on the same boat as @frodopwns, namely not wanting to use AAD but a different OIDC provider, which is what the ticket relates to.

is this possible and if so how?

**Comment by TomGeske** [comment]
No, other OIDC provider are not supported at this point.

**Comment by reallydontask** [comment]
thanks for confirming

**Comment by chrisj-platform9** [comment]
Plus 1 for this. EKS just added this https://aws.amazon.com/blogs/containers/introducing-oidc-identity-provider-authentication-amazon-eks/

**Comment by villanub** [comment]
This is something we are looking to leverage for managing a large customer base. Any update on plans to support this? EKS has taken the lead and it would be good for AKS to catch up.

**Comment by starkers** [comment]
I was thinking about migrating from AWS to Azure and just found out AKS doesn't support something a vanilla k8s cluster does..  Zoinks :/

**Comment by miwithro** [comment]
We currently have:
https://docs.microsoft.com/en-us/azure/aks/use-azure-ad-pod-identity

We are currently working on OIDC Federation support in our V2 implementation.   Looking at a Private Preview in the end of Q2 2021.

**Comment by villanub** [comment]
Feel free to reach out to me to participate in the Private Preview, happy to work on this as it is a critical piece of our requirement for success.

**Comment by ryaneorth** [comment]
@miwithro and @villanub - thanks for your update on this new feature, but that doesn't seem to address the original ask raised for this issue. Does AKS plan to support OIDC authentication to the Kubernetes API server (similar to [EKS](https://aws.amazon.com/blogs/containers/introducing-oidc-identity-provider-authentication-amazon-eks/))?

**Comment by starkers** [comment]
Yes, we just want to authorise users via an existing OIDC system, nothing fancy.. its vanilla kubernetes

**Comment by rmt** [comment]
I'm not convinced that they understand what this ticket is asking for.  This is about user identieis.

In February this year, AWS EKS introduced the much requested feature.    EKS allows you to "Associate (an OIDC) Identity Provider" with a cluster.  To configure this, you provide an Issuer URL, Client ID, Username claim and Groups claim.

For example, https://oidc.mycompany.com/, myclientID, "username", and "groups".  Kubernetes will regularly check https://oidc.mycompany.com/.well-known/openid-configuration, and from then on it will trust the username & groups in any valid JWT tokens from that provider.

The reason that this is important is that many companies manage Kubernetes clusters across multiple datacenters and clouds, and OpenID connect is _the_ standard way to do this, helping to maintain a consistent audit trail.

**Comment by keymon** [comment]
We would love to have this feature too. EKS has it and it is a game changer, to allow us to better integrate the developer workflows for our services and k8s in the same manner, without having to provide cloud credentials.

**Comment by dstockton** [comment]
GKE also has this now - https://cloud.google.com/kubernetes-engine/docs/how-to/oidc

**Comment by thomas-maurice** [comment]
Is there any updates on this ? This would be a tremendously useful feature to have.

**Comment by HerrmannHinz** [comment]
+1

**Comment by UXabre** [comment]
We also have a use-case for this, as we're using our own IDP, completely separate from AD. It would be great if we could pass a token from our API towards the kubernetes API to create resources (as this seems to be what is supported already by native kubernetes)

**Comment by miwithro** [comment]
We have it on our plans to tackle this feature early this quarter.   Once we have a firm plan, I will update this issue.

**Comment by szymonrychu** [comment]
I'm also stuck with this, other major providers AWS/GCP allow for this to happen. Currently I'm working in multicloud environment and NEED to centralize authorization around kuberneteses.
Is there any workaround for this?

**Comment by miwithro** [comment]
@xiaowanzizuibang

**Comment by UXabre** [comment]
Hi, I am correct to assume that https://github.com/Azure/AKS/issues/2861 is the follow up for this issue as it states to support external IDPs (and seems to be in the commited swimlane 🎉🎉🎉)? Perhaps this current issue can then be closed as it deals with the same topic and it would help people who are "waiting" for this feature to know that it is actually seemingly moving forward?

**Comment by CocoWang-wql** [comment]
> 
Currently AKS supports Azure AD as the identity provider for authentication. The External identity provider feature is in planning and we would post here once there is any update. https://github.com/Azure/AKS/issues/2861

**Comment by miwithro** [comment]
@atiasadir it is planned for summer.

**Comment by cre8minus1** [comment]
> @miwithro it's a middle of summer! how things are going there? :)

+1 - We are looking to standardize on OIDC across on-prem and cloud, this will really help.

**Comment by ashmuck** [comment]
Any updates on this?

**Comment by jgourmelen** [comment]
Any update ?

**Comment by CocoWang-wql** [comment]
Can you try to use[ Pinniped project](https://pinniped.dev/docs/tutorials/concierge-and-supervisor-demo/) can meet your requirements? (kindly replace the GKE steps with AKS equivalent steps)
If not, pls feel free to talk about it with me. In case you don't want to expose internal info here, you can open an Azure support ticket and attach the GitHub link, then ask the support engineer to loop me in the cal. Thank you all for your feedback.

**Comment by CocoWang-wql** [comment]
Link the[ duplicated issue](https://github.com/Azure/AKS/issues/2861) here which contains the complete conversation. Pls feel free to refer to it.

**Comment by szymonrychu** [comment]
@CocoWang-wql so the official answer is that, in Azure it's necessary to implement and maintain solution based on external operator in order to use otherwise kubernetes'es built in functionality?

Does it mean, that Azure will offer refunds for resources lost in the process (CPU, mem, monitoring, man-hours, etc)? I mean- it's an additional cost to handle such external entities in the cluster.

**Comment by miwithro** [comment]
@szymonrychu 

The Pinniped approach is currently the only option for AKS users who need this functionality.  It requires an extra "agent" on clusters, exactly in the same way as GKE's officially supported solution does.  Since Pinniped is not an officially supported solution, customers are responsible for its maintenance should they choose to use it. Many AKS features require extra pods to run on worker nodes, and, for now, the proposed approach requires that as well.  Customers are always responsible for the billing associated with these nodes.

**Comment by szymonrychu** [comment]
@miwithro yup, I can confirm, that GKE's official guide asks to deploy an operator to managed kubernetes so from that standpoint, the solution is not that different. https://cloud.google.com/kubernetes-engine/docs/how-to/oidc

What strikes me most is the difference in approach:
1. In AWS the functionality is embodied into the management of the EKS and doesn't require anything extra to be deployed (this is something I would expect)
2. In GKE it's expected to install an operator, but as I understand (e.g. based on name of the CRD), it's Google's "extension" documented, developed, tested and supported for their platform on their platform.
3. Finally in Azure, it's expected to install 3rd party's solution and be dependent on their support, good will and or payment plan.

I just think, that after all this time waiting it's not the real solution we all been expecting- especially from 2nd biggest cloud provider (or the biggest one- depends on whom you would ask).

**Comment by aslamkhan-atlan** [comment]
Is the solution still to use third party?

**Comment by PlagueHO** [comment]
Are there any updates on this? It will allow our partners to simplify and improve auth into AKS.

**Comment by cailyoung** [comment]
Hi there, with Hashicorp's recent announcement of dynamic credentials for the Kubernetes provider this would be _very very useful_ to have as a feature on our AKS clusters.

**Comment by PlagueHO** [comment]
Hi @CocoWang-wql - I'm keen to determine if this is still being planned? It is preventing 3rd party providers from implementing OIDC support for auth to AKS.

**Comment by gigabytte** [comment]
This issue is getting very long in the tooth, any updates? Are we to assume this will just work in k8s 1.30 when released for AKS?

**Comment by CocoWang-wql** [comment]
Yes, we start the design work this month. Hope to deliver it soon.

**Comment by Cryptophobia** [comment]
We waited for four years for this feature. We can wait a little more! 

At this point, it is safe to say that we probably may need to re-evaluate if anyone really needs this. In fact, lots of the users here may have decided that it is no longer required for them. Are you sure you really want to support something that is supposed to be native to Kubernetes clusters? You should know better.

Let's have a public survey at the next AKS conference on who actually needs this.

/close

**Comment by keymon** [comment]
We need it. We can live without it, and we have workarounds... but I would
rather have this feature and have the same auth than EKS.
-- 
Héctor Rivas

**Comment by shashankbarsin** [comment]
Duplicate of https://github.com/Azure/AKS/issues/2861. Closing this issue
